utils/sysfs/sysfs.go

```diff
@@ -335,25 +352,205 @@ func (fs *realSysFs) GetSystemUUID() (string, error) {
 	}
 }
 
-func (fs *realSysFs) IsCPUOnline(dir string) bool {
-	cpuPath := fmt.Sprintf("%s/online", dir)
-	content, err := ioutil.ReadFile(cpuPath)
+func (fs *realSysFs) IsCPUOnline(cpuPath string) bool {
+	return isCPUOnline(cpuPath)
+}
+
+// isCPUOnline verifies if a CPU in online. Takes path to /sys/devices/system/cpu/cpu*/ directory as a parameter.
+func isCPUOnline(cpuPath string) bool {
+	if isx86() && isZeroCPU(cpuPath) {
+		// It *is* possible to enable CPU hotplug for CPU0 but it requires setting a boot flag.
+		// It is assumed it's a rare situation.
+		return true
+	}
+
+	const isOnline = 49
+	onlinePath := filepath.Join(cpuPath, "online")
+	onlineVal, err := ioutil.ReadFile(onlinePath)
+	// First, make sure that no error occurred.
 	if err != nil {
-		pathErr, ok := err.(*os.PathError)
-		if ok {
-			if errors.Is(pathErr.Unwrap(), os.ErrNotExist) && isZeroCPU(dir) {
+		// If error occurred and is os.PathError then it might be os.ErrNotExists.
+		pathErr, isPathError := err.(*os.PathError)
+		if isPathError {
+			if !os.IsNotExist(pathErr.Unwrap()) {
+				// If online file exists but reading failed we assume that CPU is offline.
+				klog.V(2).Infof("Cannot determine CPU %s online state, assuming offline", cpuPath)
+				return false
+			}
+			// If online file does not exists then we should check one more location.
+			klog.V(2).Infof("%s not found, checking %s directory", onlinePath, filepath.Join(cpuPath, "hotplug"))
+			if isOnlineHotplug(cpuPath) {
 				return true
 			}
+			klog.V(2).Infof("unable to determine hotplug state of %s CPU; assuming that CPU if offline", cpuPath)
+			return false
 		}
-		klog.Warningf("unable to read %s: %s", cpuPath, err.Error())
+	}
+	// Check if online file value is 1.
+	onlineVal = bytes.TrimSpace(onlineVal)
+	if len(onlineVal) == 0 || onlineVal[0] != isOnline {
+		klog.V(2).Infof("CPU %s is offline", cpuPath)
+		return false
+	}
+	return true
+}
+
+func isx86() bool {
+	uts := unix.Utsname{}
+	err := uname(&uts)
+	if err != nil {
 		return false
 	}
-	trimmed := bytes.TrimSpace(content)
-	return len(trimmed) == 1 && trimmed[0] == 49
+	return uts.Machine[0] == 120 && uts.Machine[1] == 56 && uts.Machine[2] == 54
 }
 
 func isZeroCPU(dir string) bool {
 	regex := regexp.MustCompile("cpu([0-9]*)")
 	matches := regex.FindStringSubmatch(dir)
 	return len(matches) == 2 && matches[1] == "0"
 }
+
+// isOnlineHotplug checks if CPU state matches online state from /sys/devices/system/cpu/hotplug/states.
+// It allows to determine if CPU is online on some ARM platforms such as RPi4.
+// See: https://www.kernel.org/doc/html/latest/core-api/cpu_hotplug.html#testing-of-hotplug-states
+func isOnlineHotplug(cpuPath string) bool {
```

Comment:
just for my understanding, as i understand this function

1. looks at `cat /sys/devices/system/cpu/hotplug/states`, trying to find a line like

```
...
206: online
```

Then it checks `hotplug/state` on a cpu, e.g. 

```
cat /sys/devices/system/cpu/cpu0/hotplug/state
206
```

and checks if the number matches. What does "206" represent? if a cpu will be disabled is it expected the state not match anymore?

Reply:
On my RPi `206` represents online state but you mileage will vary: I bet it will be different on random x86 VM. This is why we need to parse `states` file first.

Reply:
Digging a bit deeper: take a look at [`cpuhp_setup_state()`](https://elixir.bootlin.com/linux/latest/source/include/linux/cpuhotplug.h#L208) - it allows to register callbacks for entering and leaving particular CPU state. It use used across the kernel in more than 100 places. Entries in `states` file corresponds to second argument to this function. 
Some states are "generic" and are [predefined](https://elixir.bootlin.com/linux/latest/source/kernel/cpu.c#L1510).
Does the above answer your question?

Reply:
I see interesting, thanks for the info!